### ISSUE TITLE
*? matching sometimes matches too much

### ISSUE THREAD

**Issue by kilobyte** [Genivia/ugrep#174]
Given a file:
```
foo
bar
baz
quux
bar
baz
quux
```
matching with **ug '`f.*(\n.*)*?\nq`'** erroneously grabs all lines.

Same happens with `f.*(\n.*)*?\nqu` and `f.*(\n.*)*?\nquu` — but `f.*(\n.*)*?\nquux` works correctly.

**Comment by genivia-inc** [comment]
The `*?` form is a "lazy quantifier" which may behave in unexpected ways sometimes, especially when newlines are included in the pattern as in your case. Note that the pattern matches the first part of the file all the way up to the `q` or `quux`. It is easier to view the multi-line matches with option `-n` to display line numbers where `|` indicates that multiple lines are joined together to form a single match.

**Comment by genivia-inc** [comment]
Ugrep (through RE/flex) adds lazy-quantifier capability which isn't well defined in POSIX for some non-trivial patterns such as lazy/greedy nesting. You have a greedy repeat nested inside a lazy repeat. You may want to use `'f.*(\n.*?)*?\nq'` as a pattern with a nested lazy quantifier or use option `-P`.

**Comment by genivia-inc** [comment]
Looking into the details of DFA/POSIX matching versus Perl matching for this particular case reveals some interesting differences. Note that POSIX "leftmost longest match" rule isn't the simple answer to this observation, but it is related. This rule makes POSIX inherently a bit more greedier, e.g. `foo(b|.*)` matches `foobar` fully with POSIX but only the first `foob` part of `foobar` with Perl matching. Likewise, there are subtle differences in the way lazy quantifiers work with POSIX. The inner `\n.*` gobbles up a full line "greedily", which has a stronger preference for the POSIX matcher to satisfy compared to the short `\nq` that lazily comes after to stop the matcher. But for the case that `\nquux` is specified as in `f.*(\n.*)*?\nquux`, then the greedy `\n.*` does not have priority any longer over the `\nquux` that lazily comes after to stop the matcher. In that case laziness wins over greediness. But on the other hand this is arguably counter intuitive. I will investigate this further.

**Comment by genivia-inc** [comment]
A quick update.

I plan to work on this issue to improve lazy quantifiers in POSIX mode (the default ugrep regex mode, like GNU grep).

Nobody has seriously attempted to implement lazy quantifiers in POSIX regular expressions, because POSIX sub-matching is ambiguous and ambiguity is "not allowed" to put it a bit simplistically. As I've pointed out, lazy quantifiers are possible for optionals and repetitions, but not for alternations like `|?` that cannot be lazy in POSIX.

It is provably true that POSIX lazy quantifiers cannot in all cases behave the same exact way that Perl regex lazy quantifiers work. However, there is an opportunity to get closer in similarity by reworking the RE/flex algorithm that annotates the DFA of the regular expression to match regex efficiently.

The tricky parts is when greedy and lazy quantifiers clash, either when used in nested forms (as in your example), or when used in sequence in a regex (which we can handle OK).

**Comment by genivia-inc** [comment]
I wrote the first algorithm in 2016 for [RE/flex](https://github.com/Genivia/RE-flex) to support POSIX lazy quantifier transformation to DFAs. There was no other algorithm or some (hacked) implementation publicly available that could do this. POSIX lazy quantifiers just don't exist, unlike Perl regex lazy quantifiers.

The improvements I am making are more closely aligned with Perl regex lazy forms. Perl pattern matching is depth-first using NFAs, essentially, which makes lazy quantifier matching trivial (just ignore lazy parts, then backtrack to match lazy parts). POSIX pattern matching is breadth-first using DFAs, essentially. DFAs have a speed advantage (linear time to match a pattern, i.e. never backtracks to match). So lazy quantifiers cannot simply be implemented by ignoring the lazy part and then backtrack when needed.

I am currently addressing a deficiency in my current algorithm implemented in RE/flex and ugrep that makes some regex patterns with lazy quantifiers match too much or too little. It is an elegant algorithm, but it is finicky with respect to lazy-greedy conflicts in a regex.

To illustrate the progress I've made, a simplified form of your regex is `(c[ab]*)*?cb` using `abc` only, where `c` stands for `\n` and `[ab]` stands for `.`. The improved algorithm generates a DFA that captures this lazy regex beautifully:
![image](https://github.com/Genivia/ugrep/assets/13117347/6c916fab-8a58-40db-b8c9-a81d1a454c67)
The state annotations `?` mark lazy positions in the regex string. Note that the DFA pattern ends in a final state for `cb` as should be.
